### PR TITLE
Add rsa_pss_rsae_sha256 to supported sigalgos list

### PR DESCRIPTION
**Pull request by lhuang04**
Summary:
To support RSA based cert.

Test Plan:
Checking the RSA cert could be parsed

Reviewers:
hanno.becker@arm.com,hannes.tschofenig@arm.com,junqi.wang@live.com,zhi.han@gmail.com

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: library/ssl_tls13_generic.c [hunk 1]
```diff
@@ -2272,16 +2278,26 @@ static int ssl_read_certificate_verify_parse( mbedtls_ssl_context* ssl,
        return( MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE_VERIFY );
     }
 
-    MBEDTLS_SSL_DEBUG_BUF( 3, "verify hash", verify_hash, verify_hash_len );
-
-    ret = mbedtls_pk_verify( &ssl->session_negotiate->peer_cert->pk,
+    mbedtls_pk_rsassa_pss_options* opts_ptr = NULL;
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]: Minor: Style

File: include/mbedtls/config.h [hunk 2]
```diff
@@ -1755,6 +1755,8 @@
  */
 #define MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
 
+#define MBEDTLS_RSA_PSS_RSAE_SHA256
```
Comment (hanno-becker) [inline include/mbedtls/config.h]: I don't think we need a new configuration option here, but we should rather reuse the existing configuration options for RSA from upstream Mbed TLS.

File: library/ssl_tls13_generic.c [hunk 3]
```diff
@@ -2272,16 +2278,26 @@ static int ssl_read_certificate_verify_parse( mbedtls_ssl_context* ssl,
...
+    mbedtls_pk_rsassa_pss_options opts;
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]: Minor: While we do support C99, most of the library places variable declarations at the beginning of the scope, and I'd prefer to stick that here, too.

**Review by hanno-becker — CHANGES_REQUESTED** [review]
Thanks @lhuang04 for raising this. I opened #99 to track support for RSA in the prototype.

Apart from style issues, the following needs addressing:
*  The PR introduces a new configuration option for RSA. Can we reuse existing configuration options from upstream Mbed TLS here?
* We should have tests for RSA certificates in `ssl-opt.sh` confirming that the change works as expected. Let me know if you need a hand with this - on first thought, it shouldn't be hard because you can configure the certificates on the command line, and we have certificates of various types in `tests/data_files` that you can use. Alternatively, you might run a RSA-only build, in which case the example applications will automatically pick an RSA certificate, IIRC.

File: include/mbedtls/config.h (hunk 2 above)
Comment (lhuang04) [reply include/mbedtls/config.h]: I removed the new configuration option, and uses `MBEDTLS_X509_RSASSA_PSS_SUPPORT` from upstream, let me know if you mean different existing RSA configuration.

**Comment by lhuang04** [comment]
@hanno-arm, could you take another look? I should have addressed most of your review comments.
Regarding to the test, looks like the server need more changes because we the tls13 server only supports ECDSA signature. 
Specifically at https://github.com/hannestschofenig/mbedtls/blob/tls13-prototype/library/ssl_tls13_generic.c#L1843, and https://github.com/hannestschofenig/mbedtls/blob/tls13-prototype/library/ssl_tls13_generic.c#L1720. 

Do you know any certificate in `data_files` could be used for such test? I tried `server2.crt`, but I got `MBEDTLS_ERR_RSA_INVALID_PADDING`. https://github.com/hannestschofenig/mbedtls/blob/tls13-prototype/library/rsa.c#L2221-L2222

File: library/ssl_tls13_generic.c [hunk 4]
```diff
@@ -2272,16 +2280,24 @@ static int ssl_read_certificate_verify_parse( mbedtls_ssl_context* ssl,
        return( MBEDTLS_ERR_SSL_BAD_HS_CERTIFICATE_VERIFY );
     }
 
-    MBEDTLS_SSL_DEBUG_BUF( 3, "verify hash", verify_hash, verify_hash_len );
-
-    ret = mbedtls_pk_verify( &ssl->session_negotiate->peer_cert->pk,
+    if (sig_alg == MBEDTLS_PK_RSASSA_PSS) {
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]: Nit: Style

**Review by hanno-becker — CHANGES_REQUESTED** [review]
Thank you @lhuang04 for the changes. I'm OK leaving the addition for server-support of RSA certificates to a later stage - we already have https://github.com/hannestschofenig/mbedtls/issues/99 to track it.

Could you elaborate on what you attempted to add the positive testing with an RSA certificate? From what I gather, there's no hope for a test in `ssl-opt.sh` because we don't support RSA from the server yet, but what we could do is add a test in `compat.sh` which uses an OpenSSL server using an RSA-based certificate such as `server2.crt`. Have you tried that?

**Comment by lhuang04** [comment]
@hanno-arm , I will give `compat.sh` a try.

**Comment by lhuang04** [comment]
@hanno-arm , could you take another look? I fixed the style, and updated `ssl_client2` in order to use it to test with openssl s_server.
I managed to run the following `compat.sh` successfully with some patch to turn on the configure, etc.
```
tests/compat.sh -m "tls1_3" -p "OpenSSL" -V "NO" -f "TLS_AES_128_GCM_SHA256" -t "RSA"
m->O tls1_3,no TLS_AES_128_GCM_SHA256 .................................. PASS
------------------------------------------------------------------------
PASSED (1 / 1 tests (0 skipped))
```

My patch:
```
diff --git a/include/mbedtls/config.h b/include/mbedtls/config.h
index e65e1109b..b42c4799a 100644
--- a/include/mbedtls/config.h
+++ b/include/mbedtls/config.h
@@ -2231,7 +2231,7 @@
  *
  * Comment this macro to disallow using RSASSA-PSS in certificates.
  */
-//#define MBEDTLS_X509_RSASSA_PSS_SUPPORT
+#define MBEDTLS_X509_RSASSA_PSS_SUPPORT
 
 /**
  * \def MBEDTLS_ZLIB_SUPPORT
@@ -3263,7 +3263,7 @@
  *
  * Requires: MBEDTLS_BIGNUM_C, MBEDTLS_OID_C
  */
-//#define MBEDTLS_RSA_C
+#define MBEDTLS_RSA_C
 
 /**
  * \def MBEDTLS_SHA1_C
diff --git a/tests/compat.sh b/tests/compat.sh
index d3b8d9e57..245e7bdfa 100755
--- a/tests/compat.sh
+++ b/tests/compat.sh
@@ -975,7 +975,7 @@ setup_arguments()
         O_SERVER_ARGS="$O_SERVER_ARGS -www"
     fi
 
-    M_CLIENT_ARGS="server_port=$PORT server_addr=127.0.0.1 force_version=$MODE"
+    M_CLIENT_ARGS="server_port=$PORT server_addr=localhost force_version=$MODE"
     O_CLIENT_ARGS="-connect localhost:$PORT -$MODE"
     G_CLIENT_ARGS="-p $PORT --debug 3 $G_MODE"
     G_CLIENT_PRIO="NONE:$G_PRIO_MODE:+COMP-NULL:+CURVE-ALL:+SIGN-ALL"
@@ -1112,7 +1112,7 @@ start_server() {
     log "$SERVER_CMD"
     echo "$SERVER_CMD" > $SRV_OUT
     # for servers without -www or equivalent
-    while :; do echo bla; sleep 1; done | $SERVER_CMD >> $SRV_OUT 2>&1 &
+    for i in "1 2 3 4"; do echo bla; sleep 1; done | $SERVER_CMD >> $SRV_OUT 2>&1 &
     PROCESS_ID=$!
 
     wait_server_start "$PORT" "$PROCESS_ID"
@@ -1397,6 +1397,7 @@ for VERIFY in $VERIFIES; do
                             TLS_AES_128_CCM_8_SHA256        \
                             TLS_CHACHA20_POLY1305_SHA256    \
                             "
+                        O_CIPHERS=""
                     else 
                             add_common_ciphersuites
                             add_openssl_ciphersuites
```

File: library/ssl_tls13_generic.c [hunk 5]
```diff
@@ -2538,16 +2547,29 @@ static int ssl_read_certificate_verify_parse( mbedtls_ssl_context* ssl,
...
     {
-        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_pk_verify", ret );
-        return( ret );
+      opts.mgf1_hash_id = md_alg;
+      if ( ( md_info = mbedtls_md_info_from_type( md_alg ) ) == NULL )
+      {
+        return ( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+      }
+      opts.expected_salt_len = mbedtls_md_get_size( md_info );
+      opts_ptr = &opts;
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]:
```suggestion
        opts.mgf1_hash_id = md_alg;
        if( ( md_info = mbedtls_md_info_from_type( md_alg ) ) == NULL )
        {
            return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
        }
        opts.expected_salt_len = mbedtls_md_get_size( md_info );
        opts_ptr = &opts;
```

File: library/ssl_tls13_generic.c [hunk 6]
```diff
@@ -2538,16 +2547,29 @@ static int ssl_read_certificate_verify_parse( mbedtls_ssl_context* ssl,
...
+            opts_ptr,
+            &ssl->session_negotiate->peer_cert->pk,
+            md_alg,
+            verify_hash,
+            verify_hash_len,
+            buf,
+            sig_len ) ) != 0 )
+    {
+      MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_pk_verify_ext", ret );
+      return( ret );
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]:
```suggestion
        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_pk_verify_ext", ret );
        return( ret );
```

**Comment by hanno-becker** [comment]
@lhuang04 Apologies for the delay, the change looks good. I made some smaller fixes and added tests to `ssl-opt.sh` and `all.sh`. Can you take a look and let me know if you're OK with them?

**Comment by lhuang04** [comment]
@hanno-arm , it looks good, really appreciated your help.

**Pull request merged**